### ISSUE TITLE
Add a metadata column to `add_observation`

### ISSUE THREAD

**Issue by englehardt** [mozilla/trackingprotection-tools#9]
It would be great to be able to attach metadata to the observations in a structured way, such that we have the flexibility to add additional context for some reports. I suggest we simply add an additional parameter to `DisconnectReport::add_observation` which accepts a json-formatted string.

**Comment by patjack** [comment]
@englehardt good idea! It will be easy to add support for this on our end. The metadata will be added as the 3rd (if no hash) or 4th parameter, correct? Here is an example:

`
["https://example.com", "https://example.com/example.js", "example-hash338014141577e61046892ee2b35fbdfa", '{"name":"example"}']
`

**Comment by englehardt** [comment]
@patjack what do you think about changing the entire observation field to be a list of key-value objects?

After thinking more about this, it seems funny to have a json-formatted string field within a json file. And I think it will get confusing to have variable length array records based on whether or not content hashes and metadata are attached.

An example of the output:
```json
{
    "domains": {
        "example.com": {
            "source": "August 2019 Mozilla 100k-site crawl",
            "reason": "Looks like a tracker.",
            "classification": "tracker",
            "observations": [{
                "resource_url": "https://foo.example.com/tracker.js",
                "site_url": "https://www.top-level-site.example/home.html",
                "content_hash": "example-hash338014141577e61046892ee2b35fbdfa",
                "metadata": {
                    "redirect_chain": ["a.com", "bar.example.com", "b.com"]
                }
            }, {
                "resource_url": "https://bar.example.com/also_tracker.js",
                "site_url": "https://www.other-site.example/index.html",
                "content_hash": "example-hash338014141577easdfsdfsdfsdfsafgdfgsdf",
                "metadata": {
                    "redirect_chain": ["bar.example.com", "a.com", "b.com"]
                }
            }]
        }
    }
}
```

The `resource_url` and `site_url` keys are fixed, and will always be present. The `content_hash` and `metadata` keys are optional and may not be present. `metadata` will be a dictionary which may have one of more keys. The key names are variable and should relatively self-documenting (as in the example above).

**Comment by englehardt** [comment]
See https://github.com/mozilla/trackingprotection-tools/pull/16.

**Comment by patjack** [comment]
@englehardt these proposed changes look great. I think it will add to the readability and make it easier to parse on our end. We'll go ahead and work on adding support for this and I'll follow-up when it's live.

**Comment by englehardt** [comment]
Closed by #17
